core/src/main/scala/org/bitcoins/core/protocol/dlc/compute/DLCUtil.scala

```diff
@@ -187,4 +192,101 @@ object DLCUtil {
                       tempContractId = offer.tempContractId)
   }
+
+  /** Checks that the oracles signatures given to us are correct
+    * Things we need to check
+    * 1. We have all the oracle signatures
+    * 2. The oracle signatures are for one of the contracts in the [[ContractInfo]]
+    *  @see https://github.com/bitcoin-s/bitcoin-s/issues/4032
+    */
+  def checkOracleSignaturesAgainstContract(
+      contractInfo: ContractInfo,
+      oracleSigs: Vector[OracleSignatures]): Boolean = {
+    contractInfo match {
+      case single: SingleContractInfo =>
+        checkSingleContractInfoOracleSigs(single, oracleSigs)
+      case disjoint: DisjointUnionContractInfo =>
+        val contractHasMatchingSigs: Vector[Boolean] = {
+          disjoint.contracts.map { single: SingleContractInfo =>
+            checkSingleContractInfoOracleSigs(single, oracleSigs)
+          }
+        }
+
+        //at least one disjoint union contract
+        //has to have matching signatures
+        contractHasMatchingSigs.exists(_ == true)
+    }
+  }
+
+  /** Check if the given [[SingleContractInfo]] has one [[OracleSignatures]]
+    * matches it inside of oracleSignatures.
+    */
+  private def checkSingleContractInfoOracleSigs(
+      contractInfo: SingleContractInfo,
+      oracleSignatures: Vector[OracleSignatures]): Boolean = {
+    require(oracleSignatures.nonEmpty, s"Signatures cannot be empty")
+    matchOracleSignatures(contractInfo, oracleSignatures).isDefined
+  }
+
+  /** Matches a [[SingleContractInfo]] to its oracle's signatures */
+  def matchOracleSignatures(
+      contractInfo: SingleContractInfo,
+      oracleSignatures: Vector[OracleSignatures]): Option[OracleSignatures] = {
+    matchOracleSignatures(contractInfo.announcements, oracleSignatures)
+  }
+
+  def matchOracleSignatures(
+      announcements: Vector[OracleAnnouncementTLV],
+      oracleSignatures: Vector[OracleSignatures]): Option[OracleSignatures] = {
+    val announcementNonces: Vector[Vector[SchnorrNonce]] = {
+      announcements
+        .map(_.eventTLV.nonces)
+        .map(_.vec)
+    }
+    val resultOpt = oracleSignatures.find { case oracleSignature =>
+      val oracleSigNonces: Vector[SchnorrNonce] = oracleSignature.sigs.map(_.rx)
+      announcementNonces.contains(oracleSigNonces)
+    }
+    resultOpt
+  }
+
+  /** Checks to see if the given oracle signatures and announcement have the same nonces */
+  private def matchOracleSignaturesForAnnouncements(
+      announcement: OracleAnnouncementTLV,
+      signature: OracleSignatures): Option[OracleSignatures] = {
+    matchOracleSignatures(
+      Vector(announcement),
+      Vector(signature)
+    )
+  }
+
+  /** Builds a set of oracle signatures from given announcements
+    * and attestations. This method discards attestments
+    * that do not have a matching announcement. Those attestments
+    * are not included in the returned set of [[OracleSignatures]]
+    */
+  def buildOracleSignatures(
```

Comment:
I'm a little confused as to why we need this?

Reply:
Specifically it seems like it is working too hard, I can't think of a situation where we would pass around a pool of attestments not associated with one another (i.e. all attestments from the same oracle would be together wouldn't they?)

Reply:
Idk, user copies the attestments wrong when highlighting them? 

Seems to me this

1. Consolidates matching of announcements and attestments to one place
2. Validates all user input